client-server-server/src/main/scala/zio/raft/server/RaftServer.scala

```diff
@@ -0,0 +1,587 @@
+  )
+
+  case class SessionConnection(
+    routingId: Option[RoutingId],
+    expiresAt: Instant
+  )
+
+  case class SessionMetadata(
+    capabilities: Map[String, String],
+    createdAt: Instant
+  )
+
+  case class PendingSession(
+    routingId: RoutingId,
+    nonce: Nonce,
+    capabilities: Map[String, String],
+    createdAt: Instant
+  )
+
+  /** Simplified RaftServer using pure functional state machine.
+    *
+    * Similar to RaftClient, the server uses a unified stream that merges all events. Each ServerState knows how to
+    * handle events and transition to new states.
+    */
+  class RaftServer(
+    transport: ZmqServerTransport,
+    config: ServerConfig,
+    actionQueue: Queue[ServerAction],
+    raftActionsOut: Queue[RaftAction],
+    stateRef: Ref[ServerState]
+  ):
+
+    /** Stream of actions to forward to Raft state machine.
+      */
+    def raftActions: ZStream[Any, Nothing, RaftAction] =
+      ZStream.fromQueue(raftActionsOut)
+
+    /** Submit a response from Raft back to a client.
+      */
+    def sendClientResponse(sessionId: SessionId, response: ClientResponse): UIO[Unit] =
+      actionQueue.offer(ServerAction.SendResponse(sessionId, response)).unit
+
+    /** Send server-initiated request to a client.
+      */
+    def sendServerRequest(sessionId: SessionId, request: ServerRequest): UIO[Unit] =
+      actionQueue.offer(ServerAction.SendServerRequest(sessionId, request)).unit
+
+    /** Confirm that a session has been committed by Raft.
+      */
+    def confirmSessionCreation(sessionId: SessionId): UIO[Unit] =
+      actionQueue.offer(ServerAction.SessionCreationConfirmed(sessionId)).unit
+
+    /** Notify server that this node has become the leader.
+      */
+    def stepUp(sessions: Map[SessionId, SessionMetadata]): UIO[Unit] =
+      actionQueue.offer(ServerAction.StepUp(sessions)).unit
+
+    /** Notify server that this node has stepped down from leadership.
+      */
+    def stepDown(leaderId: Option[MemberId]): UIO[Unit] =
+      actionQueue.offer(ServerAction.StepDown(leaderId)).unit
+
+  def make(bindAddress: String): ZIO[Scope & ZContext, Throwable, RaftServer] =
+    val config = ServerConfig.make(bindAddress)
+    for
+      validatedConfig <- ZIO.fromEither(ServerConfig.validated(config).left.map(new IllegalArgumentException(_)))
+
+      transport <- createTransport(validatedConfig)
+      actionQueue <- Queue.unbounded[ServerAction]
+      raftActionsOut <- Queue.unbounded[RaftAction]
+
+      // Start main loop in Follower state (not leader initially)
+      initialState = ServerState.Follower(None)
+      stateRef <- Ref.make[ServerState](initialState)
+
+      server = new RaftServer(transport, validatedConfig, actionQueue, raftActionsOut, stateRef)
+
+      _ <- startMainLoop(transport, validatedConfig, actionQueue, raftActionsOut, stateRef).forkScoped
+
+      // Register finalizer to cleanly shutdown all sessions on scope exit
+      _ <- ZIO.addFinalizer(
```

Comment:
curious, did you tell it to add the finalizer or did it came up with this by itself?

Reply:
I told him